cirq/contrib/qasm_import/_lexer.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from typing import Optional
+
+import ply.lex as lex
+
+from cirq.contrib.qasm_import.exception import QasmException
+
+
+class QasmLexer(object):
+    NATURAL_NUMBER = "NATURAL_NUMBER"
+
+    def __init__(self, qasm: str):
+        self.qasm = qasm
+        self.lex = lex.lex(object=self, debug=False)
+        self.lex.input(qasm)
```

Comment:
This is doing things that make more sense in a function than in a class initializer. Why is lexing tied to the construction of the lexer? Is this class a thing-that-lexes or is it a thing-that-has-been-lexed? Is it both? I don't think it should be both.

Reply:
The reason I did it because I was thinking about this as a one-time use object alongside with the parse. The user shouldn't have to call input multiple times and this way I don't have to think about "what if the user passes in another qasm string, how do I validate that I'm still in a valid state?". 

Another inspiration is that a lexer is nothing else than a fancy tokenizer (in python it's a [generator function](https://docs.python.org/3/library/tokenize.html), in Java it's a one-time use [class](https://docs.oracle.com/javase/7/docs/api/java/util/StringTokenizer.html)) : 

`lex.input` doesn't do the lexing - it only sets up the initial state for lexing, after which the `token()` function can stream the :

```
    def input(self, s):
        # Pull off the first character to see if s looks like a string
        c = s[:1]
        if not isinstance(c, StringTypes):
            raise ValueError('Expected a string')
        self.lexdata = s
        self.lexpos = 0
        self.lexlen = len(s)
```

If you feel strongly about it - I'm happy to rewire it, it will work that way too of course - it's just a tiny bit more code.

Reply:
I think it should be split out. Rule of thumb: for testability reasons, you generally want to avoid doing work in constructors ( see this: https://youtu.be/RlfLCWKxHJ0?t=332 ). Don't ask for stuff that allows you to make the thing you need, just directly ask for the thing you need.

Admittedly this is not always great. And we do break this rule. But it's a good rule of thumb.

Reply:
Sounds good - I refactored it.